### PR TITLE
User visible API for publishing to Kafka.

### PR DESCRIPTION
**Pull request by jcferretti**

### CODE REVIEW

**Comment by jcferretti** [comment]
This was created as a branch over Charles' branch `cpwright:cpw_publishToTable` for https://github.com/deephaven/deephaven-core/pull/1344, so it contains his same base code except for changes to `KafkaTools.java`.   I haven't done anything yet (as of me writing this comment) about python except for keeping the consume part of the refactoring working; ie, there is no produce part yet.

**Comment by jcferretti** [comment]
I have read the CLA Document and I hereby sign the CLA

File: Kafka/src/main/java/io/deephaven/kafka/publish/GenericRecordKeyOrValueSerializer.java [hunk 1]
```diff
@@ -133,7 +149,7 @@ FieldContext makeContext(final int size) {
             return new ContextImpl(size);
         }
 
-        abstract Object getFieldElement(int i, ContextImpl contextImpl);
+        abstract Object getFieldElement(int i, ChunkType inputChunk);
```
[the remark this answers is not in the document]
Comment (rcaudy) [reply Kafka/src/main/java/io/deephaven/kafka/publish/GenericRecordKeyOrValueSerializer.java:152]:
```
    private static GenericRecordFieldProcessor makeLongFieldProcessorWithInverseFactor(
            final String fieldName,
            final ColumnSource<?> chunkSource,
            final long denominator) {
        return makeGenericFieldProcessor(
                fieldName, chunkSource, (final int ii, final LongChunk<Attributes.Values> inputChunk) -> {
            final long raw = inputChunk.get(ii);
            if (raw == QueryConstants.NULL_LONG) {
                return null;
            }
            return raw / denominator;
        });
    }
 ```

[the remark this answers is not in the document]
Comment (rcaudy) [reply Kafka/src/main/java/io/deephaven/kafka/publish/GenericRecordKeyOrValueSerializer.java:152]:
```
    private static class GenericRecordFieldProcessorImpl<ChunkType extends Chunk<Attributes.Values>>
            extends GenericRecordFieldProcessor {
        private final ColumnSource<?> chunkSource;
        private final GetFieldElementFun<ChunkType> fun;

        public GenericRecordFieldProcessorImpl(final String fieldName, final ColumnSource<?> chunkSource,
                                               final GetFieldElementFun<ChunkType> fun) {
            super(fieldName);
            this.chunkSource = chunkSource;
            this.fun = fun;
        }

        class ContextImpl implements FieldContext {
            ChunkSource.GetContext getContext;

            ContextImpl(final int size) {
                getContext = chunkSource.makeGetContext(size);
            }

            @Override
            public void close() {
                getContext.close();
            }
        }

        @Override
        FieldContext makeContext(final int size) {
            return new ContextImpl(size);
        }

        @Override
        void processField(
                final FieldContext fieldContext,
                final WritableObjectChunk<GenericRecord, Attributes.Values> avroChunk,
                final OrderedKeys keys,
                final boolean previous) {
            final ContextImpl contextImpl = (ContextImpl) fieldContext;
            final ChunkType inputChunk;
            if (previous) {
                inputChunk = (ChunkType) chunkSource.getPrevChunk(contextImpl.getContext, keys);
            } else {
                inputChunk = (ChunkType) chunkSource.getChunk(contextImpl.getContext, keys);
            }

            for (int ii = 0; ii < inputChunk.size(); ++ii) {
                avroChunk.get(ii).put(fieldName, fun.get(ii, inputChunk));
            }
        }
    }

    @FunctionalInterface
    private interface GetFieldElementFun<ChunkType extends Chunk<Attributes.Values>> {
        Object get(final int ii, final ChunkType inputChunk);
    }

    private static <ChunkType extends Chunk<Attributes.Values>> GenericRecordFieldProcessor makeGenericFieldProcessor(
            final String fieldName,
            final ColumnSource<?> chunkSource,
            final GetFieldElementFun<ChunkType> fun) {
        return new GenericRecordFieldProcessorImpl<>(
                fieldName, chunkSource, fun);
    }
```

File: Kafka/src/main/java/io/deephaven/kafka/KafkaTools.java [hunk 2]
```diff
@@ -1024,60 +977,96 @@ public static Table consumeToTable(
...
-            return avroSpec.schema;
-        } else {
-            if (!kafkaConsumerProperties.containsKey(SCHEMA_SERVER_PROPERTY)) {
-                throw new IllegalArgumentException(
-                        "Avro schema name specified and schema server url propeorty " +
-                                SCHEMA_SERVER_PROPERTY + " not found.");
+    private static String[] getColumnNames(
+            final Table t,
+            final Produce.KeyOrValueSpec.Json jsonSpec) {
+        if (jsonSpec.exceptColumns == null) {
```
[the remark this answers is not in the document]
Comment (rcaudy) [reply Kafka/src/main/java/io/deephaven/kafka/KafkaTools.java]: You've fixed this, but now you probably want to validate the includes:
```
                    if (excludeColumns == null && includeColumns != null) {
                        return includeColumns;
                    }
```

[the remark this answers is not in the document]
Comment (rcaudy) [reply Kafka/src/main/java/io/deephaven/kafka/KafkaTools.java]:
```
                String[] getColumnNames(final Table t) {
                    if (excludeColumns != null && includeColumns != null) {
                        throw new IllegalArgumentException("Can't have excludeColumns and includeColumns");
                    }
                    if (excludeColumns == null && includeColumns == null) {
                        return t.getDefinition().getColumnNamesArray();
                    }
                    if (excludeColumns == null && includeColumns != null) {
                        // Validate includes
                        return includeColumns;
                    }
                    final String[] tableColumnNames = t.getDefinition().getColumnNamesArray();
                    final String[] mismatches = Arrays.stream(tableColumnNames)
                            .filter(cn -> !excludeColumns.contains(cn)).toArray(String[]::new); // Check needs fixing
                    if (mismatches.length > 0) {
                        throw new IllegalArgumentException(
                                "excludeColumns contains column names not on the table: " + mismatches);
                    }
                    return t.getDefinition().getColumnStream().map(ColumnDefinition::getName)
                            .filter(cn -> !excludeColumns.contains(cn)).toArray(String[]::new);
                }
```

File: Kafka/src/main/java/io/deephaven/kafka/KafkaTools.java [hunk 3]
```diff
@@ -661,28 +679,39 @@ DataFormat dataFormat() {
...
-                    for (final String colName : tableColumnNames) {
-                        if (excludeColumns != null && excludeColumns.contains(colName)) {
-                            continue;
+                    final Set<String> tableColumnsSet = new HashSet<>(Arrays.asList(tableColumnNames));
+                    if (includeColumns != null) {
+                        // Validate includes
+                        final List<String> missing = Arrays.stream(includeColumns).
+                                filter(cn -> !tableColumnsSet.contains(cn)).
+                                collect(Collectors.toList());
+                        for (final String col : includeColumns) {
```
Comment (rcaudy) [inline Kafka/src/main/java/io/deephaven/kafka/KafkaTools.java:695]: These 5 lines duplicate the 3 lines above, functionally.

File: Kafka/src/main/java/io/deephaven/kafka/KafkaTools.java [hunk 4]
```diff
@@ -661,28 +679,39 @@ DataFormat dataFormat() {
...
+                        return includeColumns;
                     }
-
-                    return colNames;
+                    final List<String> mismatches = excludeColumns.stream().
+                            filter(cn -> !tableColumnsSet.contains(cn)).
+                            collect(Collectors.toList());
+                    if (mismatches.size() > 0) {
+                        throw new IllegalArgumentException(
+                                "excludeColumns contains column names not on the table: " + mismatches);
```
Comment (rcaudy) [inline Kafka/src/main/java/io/deephaven/kafka/KafkaTools.java:710]: Maybe word this the same as the one above?
`"excludeColumns contains names not found in table columns: "`

File: Kafka/src/main/java/io/deephaven/kafka/publish/JsonKeyOrValueSerializer.java [hunk 5]
```diff
@@ -293,8 +293,8 @@ private JSONFieldProcessor makeToStringFieldProcessor(final String fieldName, fi
         return makeObjectFieldProcessor(
                 fieldName,
                 chunkSource,
-                (final ObjectNode node, final String childNodeFieldName, final Object raw) ->
-                        node.put(childNodeFieldName, Objects.toString(raw)));
+                (final ObjectNode node, final String childNodeFieldName, final Object raw) -> node
```
Comment (rcaudy) [inline Kafka/src/main/java/io/deephaven/kafka/publish/JsonKeyOrValueSerializer.java:296]: Spotless did a bad job here.

**Review by rcaudy — APPROVED** [review]

**Pull request merged**